File: libraries/rng/gaussian/gaussianrand.f90
function uniformrand(seed)
  implicit none

  ! -----------------------------------------------------------
  ! Park-Miller "minimal" Random number generator
  ! uniform distribution ]0,1[ . See Numerical Recipes ch. 7.0
  ! -----------------------------------------------------------

  integer,parameter :: rk=selected_real_kind(10,40) ! Double precision
  real(rk) :: uniformrand
  integer :: seed,IA,IM,IQ,IR,MASK
  real(rk) :: AM
  integer :: k
  parameter (IA=16807,IM=2147483647,AM=1.0/IM,IQ=127773,IR=2836,MASK=123459876)
  
  seed=ieor(seed,MASK)
  k=seed/IQ
  seed=IA*(seed-k*IQ)-IR*k
  if (seed < 0) seed=seed+IM
  uniformrand=AM*seed
  seed=ieor(seed,MASK)
  return
end function uniformrand


function gaussianrand(seed)
  implicit none
  
  ! ---------------------------------------------------------
  ! Random numbers with normal (Gaussian) distribution.
  ! Mean is 0 and standard deviation is 1
  ! See W.H.Press et al., Numerical Recipes 1st ed., page 203
  ! ---------------------------------------------------------
  
  integer,parameter :: rk=selected_real_kind(10,40) ! Double precision
  real(rk) :: gaussianrand
  integer :: seed
  real(rk) :: fac,v1,v2,r
  real(rk), save :: gset
  integer, save :: iset=0
  real(rk) :: uniformrand
  
  if (iset==0) then
1    v1 = 2.*uniformrand(seed)-1.
     v2 = 2.*uniformrand(seed)-1.
     r = v1*v1+v2*v2
     if (r.ge.1.) goto 1
     fac = sqrt(-2.0*log(r)/r)
     gset = v1*fac
     gaussianrand = v2*fac
     iset = 1
  else
     gaussianrand = gset
     iset = 0
  endif

  return
end

File: libraries/rng/gaussian/gaussianrand.c

double uniformrand(seed)
int *seed;

  /* ----------------------------------------------------------- */
  /* Park-Miller "minimal" Random number generator */
  /* uniform distribution ]0,1[ . See Numerical Recipes ch. 7.0 */
  /* ----------------------------------------------------------- */
{
  int IA,IM,IQ,IR,MASK;
  double AM,ans;
  int  k;

  IA=16807; IM=2147483647; AM=1.0/IM; IQ=127773; IR=2836; MASK=123459876;

  *seed= (*seed)^MASK;
  k=(*seed)/IQ;
  *seed = IA*((*seed)-k*IQ)-IR*k;
  if (*seed < 0) *seed = *seed+IM;
  ans=AM*(*seed);
  *seed= (*seed)^MASK;

  return(ans);
}



double gaussianrand(seed)
int  *seed;

  /* --------------------------------------------------------- */
  /* Random numbers with normal (Gaussian) distribution. */
  /* Mean is 0 and standard deviation is 1 */
  /* See W.H.Press et al., Numerical Recipes 1st ed., page 203 */
  /* --------------------------------------------------------- */
{
  double fac,v1,v2,r,ans;
  static double gset;
  static int iset=0;
  double uniformrand();

  if (iset==0) {
    do {
      v1 = 2.0*uniformrand(seed)-1.0;
      v2 = 2.0*uniformrand(seed)-1.0;
      r = v1*v1+v2*v2;
    } while (r>=1.0||r==0.0);
    fac = sqrt(-2.0*log(r)/r);
    gset = v1*fac;
    iset = 1;
    return(v2*fac);
  }
  else {
    iset = 0;
    return(gset);
  }
}
